PostProcess/TemporalAntiAliasing/src/TemporalAntiAliasing.cpp

```diff
@@ -155,39 +188,35 @@ void TemporalAntiAliasing::ComputeTemporalAccumulation(const RenderAttributes& R
-    const bool   ResetAccumulation = m_LastFrameIdx == ~0u || m_LastFrameIdx + 1 != FrameIndex;
-    m_LastFrameIdx                 = FrameIndex;
+    const Uint32 FrameIndex   = RenderAttribs.pPostFXContext->GetFrameDesc().Index;
+    const Uint32 CurrFrameIdx = (FrameIndex + 0) & 0x01;
+    const Uint32 PrevFrameIdx = (FrameIndex + 1) & 0x01;
 
     ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TextureCurrColor"}.Set(m_Resources[RESOURCE_IDENTIFIER_INPUT_COLOR].GetTextureSRV());
     ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TexturePrevColor"}.Set(m_Resources[RESOURCE_IDENTIFIER_ACCUMULATED_BUFFER0 + PrevFrameIdx].GetTextureSRV());
-    ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TextureMotion"}.Set(m_Resources[RESOURCE_IDENTIFIER_INPUT_MOTION_VECTORS].GetTextureSRV());
+    ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TextureCurrMotion"}.Set(m_Resources[RESOURCE_IDENTIFIER_INPUT_MOTION_VECTORS].GetTextureSRV());
+    ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TexturePrevMotion"}.Set(m_Resources[RESOURCE_IDENTIFIER_HISTORY_MOTION_BUFFER].GetTextureSRV());
     ShaderResourceVariableX{RenderTech.SRB, SHADER_TYPE_PIXEL, "g_TextureDepth"}.Set(m_Resources[RESOURCE_IDENTIFIER_INPUT_DEPTH].GetTextureSRV());
 
-    ScopedDebugGroup DebugGroup{RenderAttribs.pDeviceContext, "TemporalAccumulation"};
-
     ITextureView* pRTVs[] = {
         m_Resources[RESOURCE_IDENTIFIER_ACCUMULATED_BUFFER0 + CurrFrameIdx].GetTextureRTV(),
     };
     RenderAttribs.pDeviceContext->SetRenderTargets(_countof(pRTVs), pRTVs, nullptr, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
     RenderAttribs.pDeviceContext->SetPipelineState(RenderTech.PSO);
     RenderAttribs.pDeviceContext->CommitShaderResources(RenderTech.SRB, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
-
-    if (ResetAccumulation && m_IndexBuffer)
-    {
-        RenderAttribs.pDeviceContext->SetIndexBuffer(m_IndexBuffer, 0, RESOURCE_STATE_TRANSITION_MODE_TRANSITION);
-        RenderAttribs.pDeviceContext->DrawIndexed({3, VT_UINT32, DRAW_FLAG_VERIFY_ALL, 1, ResetAccumulation ? 3u : 0u});
-    }
-    else
-    {
-        RenderAttribs.pDeviceContext->Draw({3, DRAW_FLAG_VERIFY_ALL, 1, ResetAccumulation ? 3u : 0u});
-    }
+    RenderAttribs.pDeviceContext->Draw({3, DRAW_FLAG_VERIFY_ALL, 1});
     RenderAttribs.pDeviceContext->SetRenderTargets(0, nullptr, nullptr, RESOURCE_STATE_TRANSITION_MODE_NONE);
+
+    CopyTextureAttribs CopyAttribs;
+    CopyAttribs.pSrcTexture              = m_Resources[RESOURCE_IDENTIFIER_INPUT_MOTION_VECTORS].AsTexture();
+    CopyAttribs.pDstTexture              = m_Resources[RESOURCE_IDENTIFIER_HISTORY_MOTION_BUFFER].AsTexture();
+    CopyAttribs.SrcTextureTransitionMode = RESOURCE_STATE_TRANSITION_MODE_TRANSITION;
+    CopyAttribs.DstTextureTransitionMode = RESOURCE_STATE_TRANSITION_MODE_TRANSITION;
+    RenderAttribs.pDeviceContext->CopyTexture(CopyAttribs);
```

Comment:
Copying motion vectors internally adds overhead.
Instead, previous frame motion vector should be provided by the app, which can ping-pong between two buffers and avoid copy.

Reply:
It's pretty cheap, takes 0.05ms. But on the user side the code is simplified

Reply:
On WebGL it can be much more expensive. Besides, it will make it clearer what resources the algorithm needs.